packages/theme-language-server-common/src/documentLinks/DocumentLinksProvider.ts

```diff
@@ -91,3 +129,149 @@ function range(textDocument: TextDocument, node: { position: LiquidHtmlNode['pos
 function isLiquidString(node: LiquidHtmlNode): node is LiquidString {
   return node.type === NodeTypes.String;
 }
+
+function createDocumentLinkForTypeNode(
+  typeNode: JSONNode,
+  parentNode: LiquidRawTag,
+  textDocument: TextDocument,
+  root: URI,
+  blockType: string,
+): DocumentLink | null {
+  const startOffset = typeNode.offset;
+  const endOffset = typeNode.offset + typeNode.length;
+  const startPos = parentNode.body.position.start + startOffset;
+  const endPos = parentNode.body.position.start + endOffset;
+
+  const start = textDocument.positionAt(startPos);
+  const end = textDocument.positionAt(endPos);
+
+  return DocumentLink.create(
+    Range.create(start, end),
+    Utils.resolvePath(root, 'blocks', `${blockType}.liquid`).toString(),
+  );
+}
+
+function processPresetBlocks(
+  blocksNode: JSONNode,
+  parentNode: LiquidRawTag,
+  textDocument: TextDocument,
+  root: URI,
+): DocumentLink[] {
+  const links: DocumentLink[] = [];
+
+  if (blocksNode.type === 'object' && blocksNode.children) {
+    blocksNode.children.forEach((propertyNode) => {
+      const blockValueNode = propertyNode.children?.[1]; // The value node of the property
+      if (!blockValueNode) return;
+
+      // Check if the block has a 'name' key so we don't deeplink inline block types
+      const nameNode = findNodeAtLocation(blockValueNode, ['name']);
+      if (nameNode) {
+        return;
+      }
+
+      const typeNode = findNodeAtLocation(blockValueNode, ['type']);
+      if (typeNode && typeNode.type === 'string' && typeof typeNode.value === 'string') {
+        const blockType = typeNode.value;
+        if (blockType.startsWith('@')) {
+          return;
+        }
+
+        const link = createDocumentLinkForTypeNode(
+          typeNode,
+          parentNode,
+          textDocument,
+          root,
+          blockType,
+        );
+
+        if (link) {
+          links.push(link);
+        }
+      }
+
+      // Recursively process nested blocks
+      const nestedBlocksNode = findNodeAtLocation(blockValueNode, ['blocks']);
+      if (nestedBlocksNode) {
+        links.push(...processPresetBlocks(nestedBlocksNode, parentNode, textDocument, root));
+      }
+    });
+  } else if (blocksNode.type === 'array' && blocksNode.children) {
+    blocksNode.children.forEach((blockNode) => {
```

Comment:
One way to reduce this big function is to move the iteree code within the forEach into a standalone function definition.